src/RequestFactoryInterface.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace GuzzleHttp;
+
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Interface for generating HTTP requests.
+ */
+interface RequestFactoryInterface
+{
+
+    /**
+     * Applies any URI options to an existing URI to generate a new one.
+     *
+     * @param string $uri
+     * @param array  $config
+     *
+     * @return string
+     */
+    public function buildUri($uri, array $config);
+
+
+    /**
+     * Create a new request.
+     *
+     * The $options array is passed by reference so that any
+     * options that are applied to the request are then removed
+     * from the array to avoid them being applied twice.
+     *
+     * @param string $method
+     * @param string $uri
+     * @param array  $options
+     *
+     * @return RequestInterface
+     */
+    public function make($method, $uri = '', array &$options = []);
```

Comment:
Passing options by reference will probably lead to subtle bugs and unintended mutation.

Reply:
Can this method name be made more specific? Like createRequest?

Reply:
I've renamed it now. The tests fail if this isn't a reference, as `Client` relies on the options being mutated